src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs

```diff
@@ -56,8 +56,417 @@ private static unsafe void ThrowExceptionEncryptDecryptFail(string fullPath)
+                            return false;
+                        }
+                        if (destinationHandle.IsInvalid)
+                        {
+                            // Failure to open.
+                            return false;
+                        }
+                        madeNew = true;
+                    }
+
+                    // Read the destination file's volume's serial number.
+                    int destSerialNumber;
+                    if (Interop.Kernel32.GetVolumeInformationByHandle(sourceHandle, out _, false, &destSerialNumber, null, out int destVolumeFlags, null, 0) != 0)
+                    {
+                        throw new Exception("G");
+                        //return false;
+                    }
+
+                    // Check the destination file is on the same volume.
+                    // Note: this doesn't guarantee they're on the same volume for sure, so we have to take that into consideration later: https://devblogs.microsoft.com/oldnewthing/20170707-00/?p=96555.
+                    if (sourceSerialNumber != destSerialNumber)
+                    {
+                        throw new Exception("H");
+                        //return false;
+                    }
+
+                    // Quick sanity check to see if the destination supports the block copy operation, since it's basically free to check anyway.
+                    if ((destVolumeFlags & Interop.Kernel32.FILE_SUPPORTS_BLOCK_REFCOUNTING) == 0)
+                    {
+                        throw new Exception("I");
+                        //return false;
+                    }
+
+                    // Get the source volume path. Note: we need the real path here for symlinks also, hence openedSourceFullPath.
+                    if (Interop.Kernel32.GetVolumePathName(openedSourceFullPath, out var volumePath) != 0)
+                    {
+                        throw new Exception("K1");
+                        //return false;
+                    }
+
+                    // Read the source volume's cluster size.
+                    if (!Interop.Kernel32.GetDiskFreeSpace(volumePath!, out int sectorsPerCluster, out int bytesPerCluster, out _, out _))
+                    {
+                        throw new Exception("K");
+                        //return false;
+                    }
+                    long clusterSize = sectorsPerCluster * (long)bytesPerCluster;
+
+                    // Set file length to 0.
+                    if (!madeNew)
+                    {
+                        if (!Interop.Kernel32.SetEndOfFile(destinationHandle))
+                        {
+                            throw new Exception("O");
+                            //return false;
+                        }
+                    }
+
+                    // Ensure the destination file is not readonly.
+                    // Todo: are there other flags we need to exclude here?
```

Comment:
Does anyone know if there are more flags we need to set. Does anyone know if we should be clearing the readonly flag, presumably this should potentially actually just fail. This also probably also needs to go before setting the file length to 0 if we are going to clear readonly.